### PR TITLE
Final implementation of feature friend

### PR DESCRIPTION
**Pull request by JoeZZG**
This is the final pull request of friend, including four main features, namely accept friend, request friend, delete friend, view friend. These features are basic feature of a chat system.
Starting from requesting friend, it takes either an input of the name of whom you want to request a friend, or their UID, using try catch method to distinguish them in controller in back end. After the request is made, a notification will be sent to the friend, to make them aware of the friend request from the requester. The information of the request will be add to database to make sure the request is not lost. During the implementation, we found that the requester appears duplicately in requester list of the friend you want to request, then we add a new cleaning function to clean the duplicate data. 
After that is accepting friend, friend can can decide either add the requester as friend or reject them. As the same time, once the decision is made, a notification will be made to inform the requester if they are accepted by the friend or not. If acceptation is made, UID of both user will be stored in database to save each other as friend, and the requester UID will be deleted in databse to denote that there was a decision made. During the implementation, we found that if friend request is accepted twice, duplicate UID appears in database, which is similar to what occurs in request friend feature. We made a similar cleaner function to clean the duplicate data.
For deleting friend, it takes two type of inputs as requesting friend. Server asserts if the user name or UID is correct to make further decision. If the deletion is executed, a notification will be sent to both friend and you for awareness. 
For view friend, it does not take any input from client but the computer of client will "add" requester information and send to server. It returns back both friend`s name and their UID to client`s side. 
For code smell, the biggest problem we found was that duplicate code in DSGateway in server side. For example, getUID() or getName() appears almost each DSGateway of features. To fix it, we move some duplicate code to a common DSGateway to store some common function or method that all four features need it. Another major code smell is long parameter list, in send method between client and server, some features have to take more than five parameters. To solve it, we made a convention that send() method only takes three parameter, namely peerport, address and string of content. The first two are important so we cannot change it. For content, we originally may have more than one input. Like in view friend, it originally had a list of uid and a list of name. Then we change it to a string which takes " " as spliter and odd item is uid and even item is name to assemble the two list. 
We also followed the structural design pattern of adaptor pattern. Each of friend feature has an in put adaptor both in front end and the back end. In Abstract Factory Pattern, we used try and catch method in both request friend and delete friend to try to process two different inputs instead of if else method which may not work in this case.
For ethics part, in particular, we followed Flexibility in Use. In request friend and delete friend, both UID and name inputs are valid, which makes it possible for those who only know uid or name of the friend to complete the action. Another one is Tolerance for Error. Once a wrong UID or name is entered, a friendly feedback will be sent to inform the client.
For clean architecture in general, we strictly follow the rule of lean architecture, both in client side and server side. Each feature has their own controller, adaptor, interactor, database gateway, input and output model. In all of the four features, only interactor called the databse gateway which is a key point of clean architecture. 
The code follows SOLID principle as well. For example, Single Responsibility. Each class has only one responsibility, like that request friend can only request a friend but cannot accept friend or do other features. It makes the dependency graph less messy and coupling. Make the code more organizable. Then the test for request friend only need to assert if a new request is saved in database but needless to know if it is adding friend or other features.
The four features are mostly done by JoeZZG, with assistance from Kevin, and other team members. Hereby appreciated. Please make comments to improve the code.

### CODE REVIEW

**Review by davidlyj — DISMISSED** [review]
Changes look good. Like that some refactoring was included in this PR. Comments were also useful in helping to understand the code.

File: src/main/java/server/usecases/friendinteractors/acceptfriend/acceptFriendInteractor.java [hunk 1]
```diff
@@ -19,15 +19,28 @@ public acceptFriendInteractor(acceptFriendDSGateway dataBase, acceptFriendOutput
...
+        dataBase.cleanRequesterListDuplicateUID(friendid);
         if(Objects.equals(ifAccept, "True")){
             dataBase.acceptFriendbyID(friendid, requesterid);
-            output.success(requesterid, friendid, friendName, requesterAddress, requesterPeerPort);
+
+            String[] requesterData = dataBase.readUser(requesterid);
+            String[] friendData = dataBase.readUser(friendid);
+
+            String[] toRequester = Arrays.copyOfRange(friendData, 0, 5);
+            toRequester[3] = toRequester[3].split("-")[0];
```
Comment (xingjianll) [inline src/main/java/server/usecases/friendinteractors/acceptfriend/acceptFriendInteractor.java:37]: Create a private chat between the friend and the requester when the friend request is accepted, send private chat to both users.

File: src/main/java/server/usecases/friendinteractors/requestfriend/requestFriendInteractor.java [hunk 2]
```diff
@@ -15,41 +16,46 @@ public requestFriendInteractor(requestFriendOutputBoundary output, requestFriend
...
-//
-//            output.success(requesterid, requesterAddress, requesterPeerPort);
-//            output.reportToFriend(requesterid, requesterName, friendid, friendAddress, friendPeerPort);
+        if(b2 && b3 && b4 && b5){
+            String requesterName = model.getRequesterName();
+            String friendName = model.getFriendName();
+            int friendid = dataBase.getUIDbyUserName(friendName);
+            int requesterid = dataBase.getUIDbyUserName(requesterName);
+            String requesterAddress = dataBase.getAddress(requesterid);
+            int requesterPeerPort = dataBase.getPeerPort(requesterid);
```
Comment (xingjianll) [inline src/main/java/server/usecases/friendinteractors/requestfriend/requestFriendInteractor.java:45]: Im not sure if the users will ever request friend by their name, maybe comment it out?

**Review by xingjianll — CHANGES_REQUESTED** [review]
Very good work, see inline comments for things to add.

File: src/main/java/server/usecases/friendinteractors/requestfriend/requestFriendInteractor.java (hunk 2 above)
Comment (JoeZZG) [reply src/main/java/server/usecases/friendinteractors/requestfriend/requestFriendInteractor.java:45]: It works anyway

**Review by brendenmcfarlane — DISMISSED** [review]
Looks Good! I realized were all using 4396 everywhere, I think that's considered a "Magic Number" code smell and I've created an issue for us to eventually fix this.

**Review by xingjianll — APPROVED** [review]
GJ.

**Review by xingjianll — CHANGES_REQUESTED** [review]
why not pass test

**Review by xingjianll — DISMISSED** [review]
Approved.

**Review by as39248 — DISMISSED** [review]
approved

**Review by thmcheung — DISMISSED** [review]
I'm glad your feature works now! Your code looks clean and I cannot find any errors. Thank you for your contribution to the team! 💕

**Review by as39248 — APPROVED** [review]
lgtm

**Comment by xingjianll** [comment]
Outdated pull request.

**Pull request closed without merging**